tool/client.cc

```diff
@@ -540,24 +682,58 @@ bool Client(const std::vector<std::string> &args) {
     SSL_CTX_set_permute_extensions(ctx.get(), 1);
   }
 
+  std::string certPathFlag;
+  int verify = SSL_VERIFY_NONE;
   if (args_map.count("-root-certs") != 0) {
+    certPathFlag = "-root-certs";
+    verify = SSL_VERIFY_PEER;
+  }
+  // For the OpenSSL tool, simply specifying -CAfile does not imply verification
+  if (args_map.count("-CAfile") != 0) {
+    certPathFlag = "-CAfile";
+  }
+  if (!certPathFlag.empty()) {
     if (!SSL_CTX_load_verify_locations(
-            ctx.get(), args_map["-root-certs"].c_str(), nullptr)) {
+            ctx.get(), args_map[certPathFlag].c_str(), nullptr)) {
       fprintf(stderr, "Failed to load root certificates.\n");
       ERR_print_errors_fp(stderr);
       return false;
     }
-    SSL_CTX_set_verify(ctx.get(), SSL_VERIFY_PEER, nullptr);
   }
 
+  certPathFlag = "";
   if (args_map.count("-root-cert-dir") != 0) {
+    certPathFlag = "-root-cert-dir";
+    verify = SSL_VERIFY_PEER;
+  }
+  // For the OpenSSL tool, simply specifying -CApath does not imply verification
+  if (args_map.count("-CApath") != 0) {
+    certPathFlag = "-CApath";
+  }
+
+  if (!certPathFlag.empty()) {
     if (!SSL_CTX_load_verify_locations(
-            ctx.get(), nullptr, args_map["-root-cert-dir"].c_str())) {
+            ctx.get(), nullptr, args_map[certPathFlag].c_str())) {
```

Comment:
NP: The logic would be more robust (i.e., more likely to catch future bugs) if it used [`at`](https://en.cppreference.com/w/cpp/container/map/at) instead of [`operator[]`](https://en.cppreference.com/w/cpp/container/map/operator_at) to retrieve values.

Reply:
Hmmm my understanding is operator[] will create a new key during lookup if it doesn't exist while at will throw an exception. Since we obtain the args_map after parsing against a known list of keys, and check whether that key exists (args_map.count) before accessing the index, will this be beneficial?

Reply:
So, the problem is that it "creates a new key during lookup" -- creation of a new key is not a desired outcome and could result in a silent failure. The lookup for a non-existent key here should fail loudly. Explicit failures are always preferable to silent failures.

To clarify, I think the current code has the necessary checks to prevent unintended keys being created. My concern is more about future changes that might accidentally introduce a bug that results in invalid/misleading results for consumers due to the silent failure.

Reply:
noted, will address in follow up PR.